### PR TITLE
Ethash reloaded

### PR DESCRIPTION
**Pull request by LefterisJP**
Integrating the new ethash API into cpp-ethereum.

Using `git subtree` with `split` and `merge` in order to pull the two different directories of ethash that we use in cpp-ethereum as subtrees and keep history and contributions intact. The method is described [here](http://stackoverflow.com/questions/22334382/git-subtree-only-one-file-or-directory/22335955)

This is a work in progress. DO NOT merge yet.

### CODE REVIEW

**Comment by LefterisJP** [comment]
Note to self: Check if this PR is lost due to the merging https://github.com/ethereum/cpp-ethereum/pull/1756 since it was done on cpp-ethereum and not ethash.

File: libethash/internal.c [hunk 1]
```diff
@@ -37,264 +41,456 @@
...
-	node *nodes = (node *) cache;
-    ethash_compute_cache_nodes(nodes, params, seed);
+bool static ethash_compute_cache_nodes(
+	node* const nodes,
+	uint64_t cache_size,
+	ethash_h256_t const* seed
+)
+{
+	if (cache_size % sizeof(node) != 0) {
+		return false;
```
Comment (CJentzsch) [inline libethash/internal.c]: no braces

File: libethash/internal.c [hunk 2]
```diff
@@ -37,264 +41,456 @@
...
+	}
+	uint32_t const num_nodes = (uint32_t) (cache_size / sizeof(node));
+
+	SHA3_512(nodes[0].bytes, (uint8_t*)seed, 32);
+
+	for (uint32_t i = 1; i != num_nodes; ++i) {
```
Comment (CJentzsch) [inline libethash/internal.c]: { on next line

File: libethash/internal.c [hunk 3]
```diff
@@ -37,264 +41,456 @@
...
+		SHA3_512(nodes[i].bytes, nodes[i - 1].bytes, 64);
+	}
+
+	for (uint32_t j = 0; j != ETHASH_CACHE_ROUNDS; j++) {
```
Comment (CJentzsch) [inline libethash/internal.c]: braces on next line, a lot of braces style errors in this file, also check rest of file.

File: libethcore/Ethash.h [hunk 4]
```diff
@@ -66,7 +66,8 @@ class Ethash
 
 		h256 boundary;
 		h256 headerHash;	///< When h256() means "pause until notified a new work package is available".
-		h256 seedHash;
+		h256 seedHash; /// LTODO: IS this needed now that we use the block number instead?
```
Comment (CJentzsch) [inline libethcore/Ethash.h]: good question, you should only need one of them.

File: libethcore/EthashAux.cpp [hunk 5]
```diff
@@ -111,114 +84,101 @@ void EthashAux::killCache(h256 const& _s)
...
-		DEV_IGNORE_EXCEPTIONS(boost::filesystem::remove(oldMemoFile));
-		DEV_IGNORE_EXCEPTIONS(boost::filesystem::remove(oldMemoFile + ".info"));
+EthashAux::FullType EthashAux::full(BlockInfo const& _header)
+{
+	return full((uint64_t) _header.number);
+}
 
-		ethash_params p = params(_seedHash);
-		assert(!_dest || _dest.size() >= p.full_size);	// must be big enough.
+struct DAGChannel: public LogChannel { static const char* name(); static const int verbosity = 0; };
```
Comment (CJentzsch) [inline libethcore/EthashAux.cpp]: verbosity = 0 meaning it always prints the log right? So there is no way to disable it?
On my long time running (no maintenance) Raspberry Pi I don't want any logs.

File: libethcore/EthashAux.h [hunk 6]
```diff
@@ -26,6 +26,7 @@ namespace dev
 {
 namespace eth{
 
+
```
Comment (CJentzsch) [inline libethcore/EthashAux.h]: 2 empty lines in a row

**Comment by LefterisJP** [comment]
The ethash c code is not up for review. The review should be for the C++ code in cpp-ethereum only. Will address you coments today.

File: libethcore/Ethash.h (hunk 4 above)
Comment (gavofyork) [inline libethcore/Ethash.h]: for the work package, either this or the number - from within cpp-ethereum they are interchangeable.

**Pull request merged**